### PR TITLE
Improve typing of the union dispatcher function

### PR DESCRIPTION
**Pull request by thegedge**
## What does this PR do and why?

Closes https://github.com/mobxjs/mobx-state-tree/issues/1627

Currently, the dispatcher function for union types has two issues with the typing on the dispatcher function:

1. The input snapshot is typed as any.
2. It expects any type to be returned.

This PR improves both of these by

1. typing the input snapshot so that it's an input snapshot for the union types, and
2. it restricts the possible types returned from the dispatcher to the types specified in the union.

## Steps to validate locally

As always, this is all about the typechecking, so one would just have to play around with `types.union` and ensure it fails when you expect it to fail, and passes when you'd expect it to pass.

I've included a test to capture two key scenarios, but let me know if there should be more!

### CODE REVIEW

**Comment by coolsoftwaretyler** [comment]
Thanks @thegedge! I'll take a look soon. Will you please resolve the merge conflict? On my phone right now and GitHub is not showing precisely what's in conflict.

**Comment by thegedge** [comment]
Ahh, I think `master` on my fork wasn't up to date, so was just a conflict on the addition of the test. Fixed!

File: __tests__/core/type-system.test.ts [hunk 1]
```diff
@@ -1,5 +1,3 @@
-import { Exact } from "ts-essentials"
```
Comment (thegedge) [inline __tests__/core/type-system.test.ts:1]: I didn't realized it, but `Exact` from `ts-essentials` isn't really "exact"! It's just "if A and B are record types, the shapes of the two are the same", which is definitely not what we wanted. My bad!

(TS is hard)

**Comment by coolsoftwaretyler** [comment]
Thank you! Will take a look soon!

**Pull request merged**